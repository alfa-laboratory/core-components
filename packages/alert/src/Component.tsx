import React, { FC } from 'react';
import cn from 'classnames';

import { Plate, PlateProps } from '@alfalab/core-components-plate';

import styles from './index.module.css';

export type AlertProps = Omit<PlateProps, 'foldable' | 'defaultFolded' | 'leftAddons'>;

export const Alert: FC<AlertProps> = ({ className, title, ...restProps }) => (
    <Plate
        className={cn(styles.component, className)}
        buttonsClassName={styles.buttons}
        title={title ? <span className={styles.title}>{title}</span> : null}
        {...restProps}
    />
);
