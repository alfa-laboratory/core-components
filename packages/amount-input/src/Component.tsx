import cn from 'classnames';
import React, { useState, useEffect, useCallback, forwardRef } from 'react';
import { Input, InputProps } from '@alfalab/core-components-input';
import { FormControl } from '@alfalab/core-components-form-control';
import { THINSP, formatAmount, getCurrencySymbol } from '@alfalab/utils';
import { CurrencyCodes } from '@alfalab/data';
import { getFormatedValue, getAmountValueFromStr } from './utils';

import styles from './index.module.css';

export type AmountInputProps = Omit<InputProps, 'value' | 'onChange' | 'type'> & {
    /**
     * Денежное значение в минорных единицах
     * Значение null - значит не установлено
     */
    value?: number | null;

    /**
     * Валюта
     */
    currency?: CurrencyCodes;

    /**
     * Минорные единицы
     */
    minority?: number;

    /**
     * Жир
     */
    bold?: boolean;

    /**
     * Обработчик события изменения значения
     */
    onChange?: (
        e: React.ChangeEvent<HTMLInputElement>,
        payload: {
            /**
             * Денежное значение в минорных единицах
             * Значение null - значит не установлено
             */
            value: number | null;
            /**
             * Значение инпута
             */
            valueString: string;
        },
    ) => void;
};

/**
 * Компонент для ввода денежных значений
 */
export const AmountInput = forwardRef<HTMLInputElement, AmountInputProps>(
    (
        {
            value = null,
            minority = 100,
            currency = 'RUR',
            placeholder = `0\u2009${getCurrencySymbol(currency) || ''}`,
            bold = true,
            className,
            focusedClassName,
            dataTestId,
            clear = false,
            onChange,
            onBlur,
            onFocus,
            onClear,
            ...restProps
        },
        ref,
    ) => {
        const [focused, setFocused] = useState(false);
        const [inputValue, setInputValue] = useState<string>(
            formatAmount({
                // TODO: поддержать nullable в utils
                value: value as number,
                currency,
                minority,
            }).formated,
        );
        const filled = Boolean(inputValue || focused);

        const currencySymbol = getCurrencySymbol(currency);

        useEffect(() => {
            const currentAmountValue = getAmountValueFromStr(inputValue, minority);
            if (currentAmountValue !== value) {
                return setInputValue(
                    formatAmount({
                        // TODO: поддержать nullable в utils
                        value: value as number,
                        currency,
                        minority,
                    }).formated,
                );
            }

            return () => undefined;
            // eslint-disable-next-line react-hooks/exhaustive-deps
        }, [value, currency, minority]);

        const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
            const input = e.target;
            const enteredValue = input.value.replace(/\s/g, '').replace('.', ',');
            const isCorrectEnteredValue = /(^[0-9]{1,9}(,([0-9]+)?)?$|^\s*$)/.test(enteredValue);

            if (isCorrectEnteredValue) {
                const newFormatedValue = getFormatedValue(enteredValue, currency, minority);

                if (newFormatedValue === inputValue) {
                    const caret = input.selectionStart;
                    window.requestAnimationFrame(() => {
                        input.selectionStart = caret;
                        input.selectionEnd = caret;
                    });
                } else {
                    /**
                     * Поддержка положения каретки
                     * Поскольку при форматировании введенного значения могут появляться символы типа пробела
                     * или запятая - каретка прыгает в конец и ее необходимо ставить в правильное место
                     */

                    // Узнаем длину оригинального инпута с условием обрезания лишних символов

                    const [head, tail] = input.value.split(/\.|,/);
                    let notFormattedEnteredValueLength = head.length;
                    if (tail) {
                        notFormattedEnteredValueLength += 1; // запятая или точка
                        notFormattedEnteredValueLength += tail.slice(0, 2).length; // только 2 символа в минорной части
                    }

                    const diff = newFormatedValue.length - notFormattedEnteredValueLength;
                    const caret = (input.selectionStart as number) + diff;
                    window.requestAnimationFrame(() => {
                        input.selectionStart = caret;
                        input.selectionEnd = caret;
                    });
                }

                setInputValue(newFormatedValue);
                if (onChange) {
                    onChange(e, {
                        value: getAmountValueFromStr(newFormatedValue, minority),
                        valueString: newFormatedValue,
                    });
                }
            } else {
                // Не двигаем каретку когда вставляется невалидный символ
                const caret = (input.selectionStart as number) - 1;
                window.requestAnimationFrame(() => {
                    input.selectionStart = caret;
                    input.selectionEnd = caret;
                });
            }
        };

        const handleInputFocus = useCallback(
            (event: React.FocusEvent<HTMLInputElement>) => {
                setFocused(true);

                if (onFocus) {
                    onFocus(event);
                }
            },
            [onFocus],
        );

        const handleInputBlur = useCallback(
            (event: React.FocusEvent<HTMLInputElement>) => {
                setFocused(false);

                if (onBlur) {
                    onBlur(event);
                }
            },
            [onBlur],
        );

        const handleClear = useCallback(
            (event: React.MouseEvent<HTMLButtonElement>) => {
                setInputValue('');

                if (onClear) {
                    onClear(event);
                }
            },
            [onClear],
        );

        const [majorPart, minorPart] = inputValue.split(',');

        return (
            <div
                className={cn(styles.container, {
                    [styles.bold]: bold,
                    [styles.focused]: focused,
                    [styles.filled]: filled,
                })}
            >
                <FormControl
                    {...restProps}
                    className={cn(styles.fakeValueWithCurrencyContainer)}
                    focused={focused}
                    filled={filled}
                >
                    <div>
                        <span className={styles.majorPart}>{majorPart}</span>

                        <span className={styles.minorPartAndCurrency}>
                            {minorPart !== undefined && `,${minorPart}`}
                            {THINSP}
                            {currencySymbol}
                        </span>
                    </div>
                </FormControl>

                <Input
                    {...restProps}
                    clear={clear}
                    placeholder={placeholder}
                    value={inputValue}
                    className={cn(styles.component, className)}
                    focusedClassName={focusedClassName}
                    inputClassName={styles.input}
                    onChange={handleChange}
                    onFocus={handleInputFocus}
                    onBlur={handleInputBlur}
                    onClear={handleClear}
                    dataTestId={dataTestId}
                    ref={ref}
                />
            </div>
        );
    },
);
