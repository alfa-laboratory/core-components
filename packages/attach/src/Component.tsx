import React, {
    InputHTMLAttributes,
    ButtonHTMLAttributes,
    useState,
    useCallback,
    useRef,
} from 'react';
import cn from 'classnames';
import mergeRefs from 'react-merge-refs';
import { Button, ButtonProps } from '@alfalab/core-components-button';
import { ProgressBar } from '@alfalab/core-components-progress-bar';
import { KeyboardFocusable } from '@alfalab/core-components-keyboard-focusable';
import { AttachmentSBlackIcon, AttachmentMBlackIcon } from '@alfalab/icons-classic';
import { pluralize } from '@alfalab/utils';
import { truncateFilename } from './utils';

import styles from './index.module.css';

export type AttachProps = Omit<InputHTMLAttributes<HTMLInputElement>, 'size' | 'type' | 'value'> & {
    /**
     * Содержимое кнопки для выбора файла
     */
    buttonContent?: React.ReactNode;

    /**
     * Свойства для кнопки
     */
    buttonProps?: ButtonProps & ButtonHTMLAttributes<HTMLButtonElement>;

    /**
     * Дополнительный класс
     */
    className?: string;

    /**
     * Число символов, после которого имя файла будет обрезаться
     */
    maxFilenameLength?: number;

    /**
     * Текст для случая, когда файл не загружен
     */
    noFileText?: string;

    /**
     * Процент выполнения загрузки файла
     */
    progressBarPercent?: number;

    /**
     * Размер компонента
     */
    size?: 'xs' | 's' | 'm' | 'l';

    /**
     * Содержимое поля ввода, указанное по умолчанию. Принимает массив объектов типа File или null.
     */
    value?: File[] | null;

    /**
     * Идентификатор для систем автоматизированного тестирования
     */
    dataTestId?: string;
};

const MULTIPLE_TEXTS: [string, string, string] = ['файл', 'файла', 'файлов'];

export const Attach = React.forwardRef<HTMLInputElement, AttachProps>(
    (
        {
            size = 's',
            accept,
            buttonContent = 'Выберите файл',
            buttonProps = {},
            className,
            disabled,
            dataTestId,
            id,
            maxFilenameLength,
            multiple,
            noFileText = 'Нет файла',
            progressBarPercent,
            value,
            onChange,
            ...restProps
        },
        ref,
    ) => {
        const [files, setFiles] = useState(value || []);

        const inputRef = useRef<HTMLInputElement>(null);
        const labelRef = useRef<HTMLLabelElement>(null);
        const buttonRef = useRef<HTMLButtonElement>(null);

        const handleInputChange = useCallback(
            (event: React.ChangeEvent<HTMLInputElement>) => {
                if (event.target.files) {
                    setFiles(Array.from(event.target.files));
                }

                if (onChange) {
                    onChange(event);
                }
            },
            [onChange],
        );

        const handleButtonClick = useCallback(
            event => {
                if (labelRef.current) {
                    labelRef.current.click();
                }
                if (buttonRef.current) {
                    buttonRef.current.focus();
                }

                if (buttonProps.onClick) {
                    buttonProps.onClick(event);
                }
            },
            [buttonProps],
        );

        const handleClearClick = useCallback(() => {
            if (inputRef.current) {
                inputRef.current.value = '';
            }

            setFiles([]);
        }, []);

        const Icon = size === 'xs' ? AttachmentSBlackIcon : AttachmentMBlackIcon;

        const statusTextContent =
            files.length === 1 ? (
                truncateFilename(files[0].name, maxFilenameLength)
            ) : (
                <abbr title={files.map(file => file.name).join()}>
                    {files.length} {pluralize(files.length, ...MULTIPLE_TEXTS)}
                </abbr>
            );

        return (
            <div
                className={cn(
                    cn(
                        styles.component,
                        styles[size],
                        {
                            [styles.disabled]: disabled,
                        },
                        className,
                    ),
                )}
            >
                <Button
                    {...buttonProps}
                    size={size}
                    disabled={disabled}
                    view={(buttonProps && buttonProps.view) || 'outlined'}
                    leftAddons={
                        (buttonProps && buttonProps.leftAddons) || (
                            <Icon className={cn(styles.icon)} />
                        )
                    }
                    onClick={handleButtonClick}
                    ref={buttonRef}
                >
                    <span>{buttonContent}</span>
                </Button>
                <label className={cn(styles.label)} htmlFor={id} ref={labelRef}>
                    <input
                        {...restProps}
                        className={cn(styles.control)}
                        accept={accept}
                        disabled={disabled}
                        id={id}
                        multiple={multiple}
                        tabIndex={-1}
                        type='file'
                        onChange={handleInputChange}
                        ref={mergeRefs([ref, inputRef])}
                        data-test-id={dataTestId}
                    />
                </label>
                {files && files.length > 0 ? (
                    <div className={cn(styles.file)}>
                        <span>{statusTextContent}</span>
                        <KeyboardFocusable>
                            {(ref, focused) => (
                                <button
                                    aria-label='очистить'
                                    type='button'
                                    className={cn(styles.clear, {
                                        [styles.focused]: focused,
                                    })}
                                    onClick={handleClearClick}
                                    ref={ref}
                                />
                            )}
                        </KeyboardFocusable>
                        {progressBarPercent && (
                            <ProgressBar
                                className={cn(styles.progressBar)}
                                value={progressBarPercent}
                                view='positive'
                            />
                        )}
                    </div>
                ) : (
                    <div className={cn(styles.noFile)}>{noFileText}</div>
                )}
            </div>
        );
    },
);

/**
 * Для отображения в сторибуке
 */
Attach.defaultProps = {
    size: 's',
    buttonContent: 'Выберите файл',
    noFileText: 'Нет файла',
};
