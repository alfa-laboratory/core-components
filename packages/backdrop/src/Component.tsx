import React, { MouseEvent } from 'react';
import cn from 'classnames';
import { TransitionProps } from 'react-transition-group/Transition';
import { CSSTransition } from 'react-transition-group';

import { CSSTransitionClassNames } from 'react-transition-group/CSSTransition';
import styles from './index.module.css';

export type BackdropProps = Partial<TransitionProps> & {
    /**
     * Прозрачный бэкдроп
     */
    invisible?: boolean;

    /**
     * Управляет видимостью компонента
     */
    open: boolean;

    /**
     * Обработчик клика по бэкдропу
     */
    onClose?: (event: MouseEvent<HTMLElement>) => void;

    /**
     * Дополнительный класс
     */
    className?: string;

    /**
     * Классы анимации
     *
     * http://reactcommunity.org/react-transition-group/css-transition#CSSTransition-prop-classNames
     */
    transitionClassNames?: string | CSSTransitionClassNames;

    /**
     * Идентификатор для систем автоматизированного тестирования
     */
    dataTestId?: string;

    /**
     * Прозрачность бэкдропа
     */
    opacity?: number;
};

export const Backdrop: React.FC<BackdropProps> = ({
    className,
    open = false,
    invisible = false,
    timeout = 200,
    children,
    onClose,
    dataTestId,
    transitionClassNames = styles,
    opacity = 1,
    ...restProps
}) => {
    return (
        <CSSTransition
            timeout={timeout}
            unmountOnExit={true}
            classNames={transitionClassNames}
            in={open}
            appear={true}
            {...restProps}
        >
            <div
                aria-hidden={true}
                onClick={onClose}
                data-test-id={dataTestId}
                className={cn(styles.backdrop, className, {
                    [styles.invisible]: invisible,
                })}
                style={{ opacity }}
            >
                {children}
            </div>
        </CSSTransition>
    );
};
