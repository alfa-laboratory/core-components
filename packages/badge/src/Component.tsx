import React from 'react';
import cn from 'classnames';

import styles from './index.module.css';

export type BadgeProps = {
    /**
     * Дополнительный класс
     */
    className?: string;

    /**
     *  Вид компонента
     */
    view: 'icon' | 'count';

    /**
     * Размер компонента (только для view=count)
     */
    size?: 's' | 'm' | 'l';

    /**
     *  Видимость обводки вокруг иконки (только для view=icon)
     */
    visibleIconOutline?: boolean;

    /**
     * Контент компонента
     */
    content?: React.ReactElement | number;

    /**
     * Цветовое оформление иконки
     */
    iconColor?:
        | 'positive'
        | 'attention'
        | 'link'
        | 'negative'
        | 'tertiary'
        | 'secondary'
        | 'primary';

    /**
     * Идентификатор для систем автоматизированного тестирования
     */
    dataTestId?: string;
};

export const Badge = ({
    className,
    size = 'm',
    view,
    visibleIconOutline = false,
    content,
    iconColor,
    dataTestId,
}: BadgeProps) => {
    const isCountView = view === 'count';

    const isHidden = isCountView && typeof content === 'number' && content <= 0;
    const componentContent = isCountView && content && content >= 100 ? '99+' : content;

    return (
        <div
            className={cn(
                styles.component,
                styles[size],
                styles[view],
                iconColor && styles[iconColor],
                {
                    [styles.isHidden]: isHidden,
                    [styles.dot]: !content,
                    [styles.outline]: visibleIconOutline,
                },
                className,
            )}
            data-test-id={dataTestId}
        >
            {componentContent}
        </div>
    );
};
