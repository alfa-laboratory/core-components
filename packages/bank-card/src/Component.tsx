import React, { useCallback, MouseEvent, ReactNode, useState, ChangeEvent, useEffect } from 'react';
import cn from 'classnames';
import { MaskedInput } from '@alfalab/core-components-masked-input';
// Дождаться иконку альфы в icons-logotype
import { BankAlfaLColorIcon } from '@alfalab/icons-classic';

import { CameraMIcon } from '@alfalab/icons-glyph';

import { VisaXxlIcon, MastercardLIcon, MirXxlIcon } from '@alfalab/icons-logotype';

import styles from './index.module.css';
import { validateCardNumber } from './utils';

export type BankCardProps = {
    /**
     * Дополнительный класс
     */
    className?: string;

    /**
     * Цвет фона карты
     */
    backgroundColor?: string;

    /**
     * Иконка логотипа банка (размер L)
     */
    bankLogo?: ReactNode;

    /**
     * Лэйбл поля ввода
     */
    inputLabel?: string;

    /**
     * Значение поля ввода
     */
    value?: string;

    /**
     * Обработчик ввода
     */
    onChange?: (event: ChangeEvent<HTMLInputElement>, payload: { value: string }) => void;

    /**
     * Обработчик вызова камеры
     */
    onUsePhoto?: (event: MouseEvent<HTMLButtonElement>) => void;

    /**
     * Идентификатор для систем автоматизированного тестирования
     */
    dataTestId?: string;
};

// prettier-ignore
const cardMask = [/\d/, /\d/, /\d/, /\d/, ' ', /\d/, /\d/, /\d/, /\d/, ' ', /\d/, /\d/, /\d/, /\d/, ' ', /\d/, /\d/, /\d/, /\d/];
// prettier-ignore
const accountNumberMask = [/\d/, /\d/, /\d/, /\d/, ' ', /\d/, /\d/, /\d/, /\d/, ' ', /\d/, /\d/, /\d/, /\d/, ' ', /\d/, /\d/, /\d/, /\d/, ' ', /\d/, /\d/, /\d/, /\d/];

const getBrandIcon = (value = '') => {
    // Показываем логотип только после ввода всех цифр карты
    if (value.replace(/\s/g, '').length === 16 && validateCardNumber(value)) {
        if (value.startsWith('2')) return <MirXxlIcon />;
        if (value.startsWith('4')) return <VisaXxlIcon />;
        if (value.startsWith('5')) return <MastercardLIcon />;
        if (value.startsWith('6')) return <MastercardLIcon />;
    }
    return null;
};

export const BankCard = React.forwardRef<HTMLInputElement, BankCardProps>(
    (
        {
            bankLogo = <BankAlfaLColorIcon />,
            backgroundColor = '#EF3124',
            inputLabel = 'Номер карты или счёта',
            value,
            className,
            onUsePhoto,
            onChange,
            dataTestId,
        },
        ref,
    ) => {
        const uncontrolled = value === undefined;

        const [brandIcon, setBrandIcon] = useState<ReactNode>(getBrandIcon(value));

        const getMask = (newValue: string) =>
            newValue.length <= cardMask.length ? cardMask : accountNumberMask;

        const handleInputChange = useCallback(
            (event: ChangeEvent<HTMLInputElement>, payload) => {
                if (uncontrolled) {
                    setBrandIcon(getBrandIcon(event.target.value));
                }

                if (onChange) {
                    onChange(event, payload);
                }
            },
            [onChange, uncontrolled],
        );

        const renderRightAddons = useCallback(
            () => (
                <button type='button' className={styles.usePhoto} onClick={onUsePhoto}>
                    <CameraMIcon />
                </button>
            ),
            [onUsePhoto],
        );

        useEffect(() => {
            setBrandIcon(getBrandIcon(value));
        }, [value]);

        return (
            <div className={cn(styles.component, className)}>
                <div className={styles.aspectRatioContainer}>
                    <div className={styles.content} style={{ backgroundColor }}>
                        <div className={styles.bankLogo}>{bankLogo}</div>

                        <MaskedInput
                            ref={ref}
                            value={value}
                            mask={getMask}
                            block={true}
                            label={inputLabel}
                            rightAddons={renderRightAddons()}
                            inputClassName={styles.input}
                            labelClassName={styles.label}
                            filledClassName={styles.filled}
                            focusedClassName={styles.focused}
                            onChange={handleInputChange}
                            dataTestId={dataTestId}
                            inputMode='numeric'
                            pattern='[0-9]*'
                        />

                        {brandIcon && <div className={styles.brandLogo}>{brandIcon}</div>}
                    </div>
                </div>
            </div>
        );
    },
);

BankCard.defaultProps = {
    bankLogo: <BankAlfaLColorIcon />,
    backgroundColor: '#EF3124',
    inputLabel: 'Номер карты или счёта',
};
