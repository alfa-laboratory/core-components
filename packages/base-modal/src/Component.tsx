/* eslint-disable jsx-a11y/no-noninteractive-element-interactions */
import React, {
    useCallback,
    useRef,
    useState,
    forwardRef,
    MouseEvent,
    KeyboardEvent,
    ReactNode,
    useEffect,
    useMemo,
    Ref,
} from 'react';
import cn from 'classnames';
import mergeRefs from 'react-merge-refs';
import { ResizeObserver } from 'resize-observer';
import { CSSTransition } from 'react-transition-group';
import { TransitionProps } from 'react-transition-group/Transition';
import FocusLock from 'react-focus-lock';

import { Portal, PortalProps } from '@alfalab/core-components-portal';
import { Backdrop, BackdropProps } from '@alfalab/core-components-backdrop';

import { handleContainer, hasScrollbar, isScrolledToBottom, isScrolledToTop } from './utils';

import styles from './index.module.css';

export type BaseModalProps = {
    /**
     * Контент
     */
    children?: ReactNode;

    /**
     * Свойства для Бэкдропа
     */
    backdropProps?: Partial<BackdropProps> & Record<string, unknown>;

    /**
     * Нода, компонент или функция возвращающая их
     *
     * Контейнер к которому будут добавляться порталы
     */
    container?: PortalProps['getPortalContainer'];

    /**
     * Отключает автоматический перевод фокуса на модалку при открытии
     * @default false
     */
    disableAutoFocus?: boolean;

    /**
     * Отключает ловушку фокуса
     * @default false
     */
    disableFocusLock?: boolean;

    /**
     * Отключает восстановление фокуса на предыдущем элементе после закрытия модалки
     * @default false
     */
    disableRestoreFocus?: boolean;

    /**
     * Отключает вызов `callback` при нажатии Escape
     * @default false
     */
    disableEscapeKeyDown?: boolean;

    /**
     * Отключает вызов `callback` при клике на бэкдроп
     * @default false
     */
    disableBackdropClick?: boolean;

    /**
     * Содержимое модалки всегда в DOM
     * @default false
     */
    keepMounted?: boolean;

    /**
     * Управление видимостью модалки
     */
    open: boolean;

    /**
     * Дополнительный класс
     */
    className?: string;

    /**
     * Дополнительный класс
     */
    contentClassName?: string;

    /**
     * Дополнительный класс для обертки (Modal)
     */
    wrapperClassName?: string;

    scrollHandler?: 'wrapper' | 'content';

    /**
     * Пропсы для анимации (CSSTransition)
     */
    transitionProps?: Partial<TransitionProps>;

    /**
     * Обработчик события нажатия на бэкдроп
     */
    onBackdropClick?: (event: MouseEvent) => void;

    /**
     * Обработчик события нажатия на Escape
     *
     * Если `disableEscapeKeyDown` - false и модальное окно в фокусе
     */
    onEscapeKeyDown?: (event: KeyboardEvent) => void;

    /**
     * Обработчик закрытия
     */
    onClose?: (
        event: MouseEvent<HTMLElement> | KeyboardEvent<HTMLElement>,
        reason?: 'backdropClick' | 'escapeKeyDown' | 'closerClick',
    ) => void;

    /**
     * Обработчик события onEntered компонента Transition
     */
    onMount?: () => void;

    /**
     * Обработчик события onExited компонента Transition
     */
    onUnmount?: () => void;

    /**
     * Идентификатор для систем автоматизированного тестирования
     */
    dataTestId?: string;
};

export type BaseModalContext = {
    hasFooter?: boolean;
    hasHeader?: boolean;
    hasScroll?: boolean;
    headerHighlighted?: boolean;
    footerHighlighted?: boolean;
    contentRef: Ref<HTMLElement>;
    setHasHeader: (exists: boolean) => void;
    setHasFooter: (exists: boolean) => void;
    onClose: Required<BaseModalProps>['onClose'];
};

export const BaseModalContext = React.createContext<BaseModalContext>({
    hasFooter: false,
    hasHeader: false,
    hasScroll: false,
    headerHighlighted: false,
    footerHighlighted: false,
    contentRef: () => null,
    setHasHeader: () => null,
    setHasFooter: () => null,
    onClose: () => null,
});

export const BaseModal = forwardRef<HTMLDivElement, BaseModalProps>(
    (
        {
            open,
            container,
            children,
            scrollHandler = 'wrapper',
            backdropProps = {},
            transitionProps = {},
            disableBackdropClick,
            disableAutoFocus = false,
            disableFocusLock = false,
            disableEscapeKeyDown = false,
            disableRestoreFocus = false,
            keepMounted = false,
            className,
            contentClassName,
            wrapperClassName,
            onBackdropClick,
            onClose,
            onEscapeKeyDown,
            onMount,
            onUnmount,
            dataTestId,
        },
        ref,
    ) => {
        const [exited, setExited] = useState(!open);
        const [hasScroll, setHasScroll] = useState(false);
        const [hasHeader, setHasHeader] = useState(false);
        const [hasFooter, setHasFooter] = useState(false);
        const [headerHighlighted, setHeaderHighlighted] = useState(false);
        const [footerHighlighted, setFooterHighlighted] = useState(false);

        const componentRef = useRef<HTMLDivElement>(null);
        const contentRef = useRef<HTMLDivElement>(null);
        const wrapperRef = useRef<HTMLDivElement>(null);
        const scrollableNodeRef = useRef<HTMLDivElement | null>(null);
        const restoreContainerStyles = useRef<null | Function>(null);

        const shouldRender = keepMounted || open || !exited;

        const resizeObserver = useMemo(() => {
            return new ResizeObserver(() => {
                if (scrollableNodeRef.current) {
                    const scrollExists = hasScrollbar(scrollableNodeRef.current);
                    setFooterHighlighted(scrollExists);
                    setHasScroll(scrollExists);
                }
            });
        }, []);

        const addResizeHandle = useCallback(() => {
            if (scrollableNodeRef.current && contentRef.current) {
                resizeObserver.observe(scrollableNodeRef.current);
                resizeObserver.observe(contentRef.current);
            }
        }, [resizeObserver]);

        const removeResizeHandle = useCallback(() => {
            resizeObserver.disconnect();
        }, [resizeObserver]);

        const handleScroll = useCallback(() => {
            if (!scrollableNodeRef.current) return;

            if (componentRef.current) {
                if (hasHeader) {
                    setHeaderHighlighted(
                        isScrolledToTop(scrollableNodeRef.current) === false &&
                            componentRef.current.getBoundingClientRect().top <= 0,
                    );
                }

                if (hasFooter) {
                    setFooterHighlighted(
                        isScrolledToBottom(scrollableNodeRef.current) === false &&
                            componentRef.current.getBoundingClientRect().bottom >=
                                window.innerHeight,
                    );
                }
            }
        }, [hasFooter, hasHeader]);

        const handleClose = useCallback<Required<BaseModalProps>['onClose']>(
            (event, reason) => {
                if (onClose) {
                    onClose(event, reason);
                }

                if (reason === 'backdropClick' && onBackdropClick) {
                    onBackdropClick(event as MouseEvent);
                }

                if (reason === 'escapeKeyDown' && onEscapeKeyDown) {
                    onEscapeKeyDown(event as KeyboardEvent);
                }

                return null;
            },
            [onBackdropClick, onClose, onEscapeKeyDown],
        );

        const handleBackdropClick = (event: MouseEvent<HTMLElement>) => {
            if (!disableBackdropClick) {
                handleClose(event, 'backdropClick');
            }
        };

        const handleKeyDown = useCallback(
            (event: KeyboardEvent<HTMLDivElement>) => {
                /*
                 * Чтобы сохранить дефолтное поведение элементов и событий форм,
                 * обработчик не устанавливает event.preventDefault()
                 */
                if (event.key !== 'Escape') {
                    return;
                }

                // Если есть обработчик escape на body
                event.stopPropagation();

                if (!disableEscapeKeyDown && handleClose) {
                    handleClose(event, 'escapeKeyDown');
                }
            },
            [disableEscapeKeyDown, handleClose],
        );

        const handleEntered: Required<TransitionProps>['onEntered'] = useCallback(
            (node, isAppearing) => {
                scrollableNodeRef.current =
                    scrollHandler === 'content' ? componentRef.current : wrapperRef.current;

                addResizeHandle();

                if (scrollableNodeRef.current) {
                    scrollableNodeRef.current.addEventListener('scroll', handleScroll);
                    handleScroll();
                }

                if (transitionProps.onEntered) {
                    transitionProps.onEntered(node, isAppearing);
                }

                if (onMount) onMount();
            },
            [addResizeHandle, handleScroll, onMount, scrollHandler, transitionProps],
        );

        const handleExited: Required<TransitionProps>['onExited'] = useCallback(
            node => {
                removeResizeHandle();

                setExited(true);

                if (scrollableNodeRef.current) {
                    scrollableNodeRef.current.removeEventListener('scroll', handleScroll);
                }

                if (transitionProps.onExited) {
                    transitionProps.onExited(node);
                }

                if (onUnmount) onUnmount();
            },
            [handleScroll, onUnmount, removeResizeHandle, transitionProps],
        );

        useEffect(() => {
            if (open) {
                restoreContainerStyles.current = handleContainer(
                    (container ? container() : document.body) as HTMLElement,
                );
            }

            return () => {
                if (restoreContainerStyles.current) {
                    restoreContainerStyles.current();
                    restoreContainerStyles.current = null;
                }
            };
            // eslint-disable-next-line react-hooks/exhaustive-deps
        }, [open]);

        useEffect(() => {
            if (open) setExited(false);
        }, [open]);

        useEffect(() => {
            return () => {
                resizeObserver.disconnect();
            };
            // eslint-disable-next-line react-hooks/exhaustive-deps
        }, []);

        const contextValue = useMemo<BaseModalContext>(
            () => ({
                hasHeader,
                hasFooter,
                hasScroll,
                headerHighlighted,
                footerHighlighted,
                contentRef,
                setHasHeader,
                setHasFooter,
                onClose: handleClose,
            }),
            [hasHeader, hasFooter, hasScroll, headerHighlighted, footerHighlighted, handleClose],
        );

        if (!shouldRender) return null;

        return (
            <Portal getPortalContainer={container}>
                <BaseModalContext.Provider value={contextValue}>
                    <FocusLock
                        autoFocus={!disableAutoFocus}
                        disabled={disableFocusLock || !open}
                        returnFocus={!disableRestoreFocus}
                    >
                        <div
                            role='dialog'
                            className={cn(styles.wrapper, wrapperClassName, {
                                [styles.hidden]: !open && exited,
                            })}
                            ref={mergeRefs([ref, wrapperRef])}
                            onKeyDown={handleKeyDown}
                            tabIndex={-1}
                            data-test-id={dataTestId}
                        >
                            {Backdrop && (
                                <Backdrop
                                    {...backdropProps}
                                    open={open}
                                    onClick={handleBackdropClick}
                                />
                            )}
                            <CSSTransition
                                appear={true}
                                timeout={200}
                                classNames={styles}
                                {...transitionProps}
                                in={open}
                                onEntered={handleEntered}
                                onExited={handleExited}
                            >
                                <div className={cn(styles.component, className)} ref={componentRef}>
                                    <div className={cn(styles.content, contentClassName)}>
                                        {children}
                                    </div>
                                </div>
                            </CSSTransition>
                        </div>
                    </FocusLock>
                </BaseModalContext.Provider>
            </Portal>
        );
    },
);
