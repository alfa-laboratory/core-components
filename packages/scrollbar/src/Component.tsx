import React, { useRef, useEffect, forwardRef, MouseEvent, ReactNode } from 'react';
import mergeRefs from 'react-merge-refs';
import cn from 'classnames';
import PerfectScrollbar from 'perfect-scrollbar';

import styles from './index.module.css';

export type ScrollbarProps = {
    /**
     * Дополнительный класс
     */
    className?: string;

    /**
     * Идентификатор для систем автоматизированного тестирования
     */
    dataTestId?: string;

    /**
     * Контент
     */
    children?: ReactNode;

    /**
     * Обработчик скрола
     */
    onScroll?: (event: MouseEvent<HTMLDivElement>) => void;
};

export const Scrollbar = forwardRef<HTMLDivElement, ScrollbarProps>(
    ({ className, dataTestId, children, onScroll, ...restProps }, ref) => {
        const containerRef = useRef<HTMLDivElement>(null);

        useEffect(() => {
            let ps: PerfectScrollbar;

            if (containerRef.current) {
                ps = new PerfectScrollbar(containerRef.current, {});
            }

            return () => {
                if (ps) ps.destroy();
            };
        }, []);

        return (
            <div
                ref={mergeRefs([ref, containerRef])}
                className={cn(className, styles.component)}
                onScroll={onScroll}
                data-test-id={dataTestId}
                {...restProps}
            >
                {children}
            </div>
        );
    },
);
