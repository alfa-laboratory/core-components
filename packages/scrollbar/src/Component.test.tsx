import React from 'react';
import { fireEvent, render } from '@testing-library/react';

import { Scrollbar } from './index';
import { ScrollbarProps } from './Component';

const renderScrollbar = (props?: ScrollbarProps) => (
    <Scrollbar {...props}>
        <div
            style={{
                height: '200px',
            }}
        />
    </Scrollbar>
);

describe('Scrollbar', () => {
    describe('Display tests', () => {
        it('should display correctly', () => {
            const dataTestId = 'test-id';

            const { container } = render(renderScrollbar({ dataTestId }));

            expect(container).toMatchSnapshot();
        });
    });

    it('should call onScroll', async () => {
        const onScroll = jest.fn();

        const dataTestId = 'test-id';

        const { getByTestId } = render(renderScrollbar({ onScroll, dataTestId }));

        fireEvent.scroll(getByTestId(dataTestId));

        expect(onScroll).toBeCalledTimes(1);
    });
});
