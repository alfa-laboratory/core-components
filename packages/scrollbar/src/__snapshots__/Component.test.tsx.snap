// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scrollbar Display tests should display correctly 1`] = `
<div>
  <div
    class="component ps"
    data-test-id="test-id"
  >
    <div
      style="height: 200px;"
    />
    <div
      class="ps__rail-x"
      style="left: 0px; top: 0px;"
    >
      <div
        class="ps__thumb-x"
        style="left: 0px; width: 0px;"
        tabindex="0"
      />
    </div>
    <div
      class="ps__rail-y"
      style="top: 0px; left: 0px;"
    >
      <div
        class="ps__thumb-y"
        style="top: 0px; height: 0px;"
        tabindex="0"
      />
    </div>
  </div>
</div>
`;
